tools/bigquery-zos-mainframe-connector/gszutil/src/main/scala/com/google/cloud/bqsh/cmd/Extract.scala

```diff
@@ -0,0 +1,151 @@
+    logger.info(s"Starting $name\n$cfg")
+    val bq: BigQuery = Services.bigQuery(cfg.projectId, cfg.location, creds)
+
+    try {
+      if (cfg.sourceTable.isEmpty) {
+        // Extract results of a SQL query
+        val query =
+          if (cfg.sql.nonEmpty) cfg.sql
+          else {
+            cfg.dsn match {
+              case Some(dsn) =>
+                logger.info(s"Reading query from DSN: $dsn")
+                zos.readDSNLines(dsn).mkString("\n")
+              case None =>
+                logger.info("Reading query from DD: QUERY")
+                zos.readDDString("QUERY", "\n")
+            }
+          }
+        logger.info(s"SQL Query:\n$query")
+        if (query.isEmpty) {
+          val msg = "Empty extract query"
+          logger.error(msg)
+          return Result.Failure(msg)
+        }
+
+        val queryConfig = QueryConfig(sql = query, queryDSN = cfg.queryDSN, timeoutMinutes = cfg.timeoutMinutes,
+          allowLargeResults = cfg.allowLargeResults, destinationTable = cfg.destinationTable, dryRun = cfg.dryRun,
+          maximumBytesBilled = cfg.maximumBytesBilled, requireCache = cfg.requireCache, useCache = cfg.useCache,
+          useLegacySql = cfg.useLegacySql, projectId = cfg.projectId, datasetId = cfg.datasetId,
+          location = cfg.location, debugMode = cfg.debugMode, batch = cfg.batch, sync = true,
+          jobProperties = cfg.jobProperties, replace = cfg.replace)
+        val (jobId, job) = BqQueryJobExecutor(bq, queryConfig, zos).execute(query, Query.configureQueryJob)
+        val table = job.getConfiguration[QueryJobConfiguration].getDestinationTable
+        val rowCount = BQ.rowCount(bq, table)
+        val stats = job.getStatistics[QueryStatistics]
+        val result = extract(table, bq, cfg, zos)
+
+        // Publish results
+        if (cfg.statsTable.nonEmpty) {
+          val statsTable = BQ.resolveTableSpec(cfg.statsTable, cfg.projectId, cfg.datasetId)
+          val jobId = BQ.genJobId(cfg.projectId, cfg.location, zos, "extract")
+          val tblspec = s"${statsTable.getProject}:${statsTable.getDataset}.${statsTable.getTable}"
+          logger.debug(s"Writing stats to $tblspec, with jobId ${jobId}")
+          StatsUtil.retryableInsertJobStats(zos, jobId, bq, statsTable, jobType = "export", recordsOut = result.activityCount)
+        }
+
+        logger.info(s"Extract completed, extracted=${rowCount} rows, " +
+          s"time took: ${(System.currentTimeMillis() - startTime) / 1000} seconds.")
+        result
+      } else {
+        // Extract contents of a table
+        val table: TableId = BQ.resolveTableSpec(cfg.sourceTable, cfg.projectId, cfg.datasetId)
+        logger.info(s"Extract from source table: $table")
+        val rowCount = BQ.rowCount(bq, table)
+        val result = extract(table, bq, cfg, zos)
+
+        // Publish results
+        if (cfg.statsTable.nonEmpty) {
+          val statsTable = BQ.resolveTableSpec(cfg.statsTable, cfg.projectId, cfg.datasetId)
```

Comment:
same comment as in https://github.com/GoogleCloudPlatform/professional-services/pull/902#discussion_r990410420